### PR TITLE
chore(ci): Setup GitHub Actions

### PR DESCRIPTION
**Pull request by sidvishnoi**
Closes https://github.com/w3c/respec/issues/2481
So we can either run some steps in parallel in different jobs and repeat a lot of work, or run them in sequence.

Parallel:
**Con**: Repeats _"checkout, node install, npm install + prepare + build"_, which takes nearly 1m 20s. That would be a waste of CPU cycles.
**Pros**: Probably faster overall most likely. Want me to test?

Series:
**Pros**: If we run in series, we reuse data from previous steps.
**Cons**: Most likely slower.
In total, it took ~5m, which is still much faster than Travis (8-10m)

Also, Firefox tests ran twice as slow (2m), compared to Chrome (55s).

---

| | Strategy |  Time |
| --- | --- | --- |
| Series | `Lint -> headless -> karmaChrome -> karmaFF` |  5m |
| Travis | `Lint -> headless -> (...karma)` |  8-10m |
| Travis-like | `Lint -> headless -> (...karma)` |  6m 16s |
| Parallel | `(lint, headless, ...karma)` |  3m 23s |
| Lint, then test | `Lint -> (headless, ...karma)` |  4m 43s |

### CODE REVIEW

**Comment by sidvishnoi** [comment]
@marcoscaceres Why were GH_API_KEY and GH_USER environment variables needed on Travis? I can't find them being used anywhere.

File: .github/workflows/pr-push.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
+name: Pull Request
```
Comment (sidvishnoi) [inline .github/workflows/pr-push.yml]: Needs a better name

**Comment by sidvishnoi** [comment]
So we can either run some steps in parallel in different jobs and repeat a lot of work, or run them in sequence.

Parallel:
**Con**: Repeats _"checkout, node install, npm install + prepare + build"_, which takes nearly 1m 20s. That would be a waste of CPU cycles.
**Pros**: Probably faster overall most likely. Want me to test?

Series:
**Pros**: If we run in series, we reuse data from previous steps.
**Cons**: Most likely slower.
In total, it took ~5m, which is still much faster than Travis - 8m 11s.

Also, Firefox tests ran twice as slow (2m), compared to Chrome (55s).

File: .github/workflows/pr-push.yml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+on: pull_request
+
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v1
```
Comment (sidvishnoi) [inline .github/workflows/pr-push.yml]: TODO (if decide to run parallel jobs): reuse nodes in yml

Comment (marcoscaceres) [reply .github/workflows/pr-push.yml]: Previously we ran lint first, then headless, then the browser tests in parallel. Is that still doable? 

Also, yes, Firefox is 2x slower than Chrome unfortunately. Known issue.

Comment (sidvishnoi) [reply .github/workflows/pr-push.yml]: I think we should be able to do something like:
``` yaml
jobs:
  lint: ... # 1st
  headless:
   needs: lint # 2nd
  tests1:
    needs: headless # 3.1, parallel
  tests2:
    needs: headless # 3.2, parallel
```
Each job will require to repeat the "clone, install and build" steps though (which takes 1min at least).

But why would it be better than running all in parallel? Early exit?

Comment (marcoscaceres) [reply .github/workflows/pr-push.yml]: Lint is usually really quick, same with headless. The karma jobs were usually really slow.

Comment (sidvishnoi) [reply .github/workflows/pr-push.yml]: The lint step took 15 seconds, but set up step for lint took 45 seconds (no build needed after install).
In headless, tests took 18s, but setup took 1m 30s.
If we wait for headless and lint, we've added >2 min of delay.

If we run in series, that 1m 30s is reused in all steps (clone, install and build once, then do other steps - but other steps are done in series).
Ideally, I would've loved if the specific steps ran in parallel after common steps (not possible right now).

Comment (sidvishnoi) [reply .github/workflows/pr-push.yml]: Decisions decisions

Comment (marcoscaceres) [reply .github/workflows/pr-push.yml]: I think we should mirror what we originally had, even if it adds 2 minutes. The linting is usually what catches most issues for us, and there is no point in running karma if eslint fails.

Comment (sidvishnoi) [reply .github/workflows/pr-push.yml]: | Test  | Time |
| --- | --- |
| Lint | 1m 04s |
| Headless | 1m 47s |
| Karma Chrome | 2m 33s |
| Karma Firefox | 3m 25s |
| **Total** | 6m 16s |

**Summary**:

| Strategy |  Time |
| --- | --- |
| Parallel |  3m 23s |
| Travis like |  6m 16s |
| Series |  5m |
| Travis |  8-10m |

Comment (sidvishnoi) [reply .github/workflows/pr-push.yml]: I think we better run series... looking at the results in PR description

Comment (marcoscaceres) [reply .github/workflows/pr-push.yml]: Whatever you think it’s best. I don’t mind.

**Comment by sidvishnoi** [comment]
@marcoscaceres @saschanaz Please see https://github.com/w3c/respec/pull/2482#issuecomment-523994974
Also, do we need to add `SNYK_TOKEN` during `npm install`? If so, @marcoscaceres  can you get it added to repository secrets?

**Comment by marcoscaceres** [comment]
I don’t think we need the snyk token. But I can add it if anything breaks.

**Comment by saschanaz** [comment]
> @marcoscaceres Why were GH_API_KEY and GH_USER environment variables needed on Travis? I can't find them being used anywhere.

Maybe just some old remnants? 🤔

**Comment by saschanaz** [comment]
How did you mark things as Required, BTW? And what are the five checks that never runs?

image

**Comment by sidvishnoi** [comment]
> How did you mark things as Required, BTW?

I did nothing. Looks automatic.

> what are the five checks that never runs?

Not sure where they came from. Didn't see them before. Looks something from past. 
Lets try merging and see if they appear on other PRs.

**Comment by saschanaz** [comment]
> > How did you mark things as Required, BTW?
> 
> I did nothing. Looks automatic.
> 
> > what are the five checks that never runs?
> 
> Not sure where they came from. Didn't see them before. Looks something from past.
> Lets try merging and see if they appear on other PRs.

#2484 is already blocked by them 🤔 Hmm...

**Comment by sidvishnoi** [comment]
Well, that probably tells why GitHub actions are in beta right now 😄

File: .github/workflows/pr-push.yml [hunk 3]
```diff
@@ -0,0 +1,61 @@
+name: Node CI
+on:
+  push:
+    branches:
+      - develop
+      - gh-pages
+  pull_request: {}
```
Comment (saschanaz) [inline .github/workflows/pr-push.yml]: This is triggered [whenever a PR changes](https://help.github.com/en/articles/events-that-trigger-workflows#webhook-events):

>Triggered when a pull request is assigned, unassigned, labeled, unlabeled, opened, edited, closed, reopened, synchronize, ready_for_review, locked, unlocked or when a pull request review is requested or removed. See below to learn how this event works with forked repositories.

Maybe just do `push` without `branches` limit?

Comment (sidvishnoi) [reply .github/workflows/pr-push.yml]: I did test if it got triggered on changes mentioned there (I assigned myself this PR) and nothing happened. Probably a documentation error.

**Comment by sidvishnoi** [comment]
@marcoscaceres Lets merge this and see what happens?

**Comment by marcoscaceres** [comment]
YOLO!

**Comment by saschanaz** [comment]
@marcoscaceres please do a force merge, because something is broken now.

**Comment by sidvishnoi** [comment]
@marcoscaceres Can you merge this? We don't have the permissions to force merge 😸

**Comment by marcoscaceres** [comment]
Boom!

**Pull request merged**